FileIOUtils.cpp

```diff
@@ -0,0 +1,426 @@
+/// </param>
+/// <returns>
+/// True if the save was successful.
+/// </returns>
+bool FileIOUtils::ImportTile8x8GfxData(QWidget *parent, QVector<QRgb> ref_palette, std::function<void (QByteArray&, QWidget *)> TilesReplaceCallback)
+{
+    // Load gfx bin file
+    QString fileName = QFileDialog::getOpenFileName(parent,
+                                                    QObject::tr("Load Tileset graphic bin file"), QString(""),
+                                                    QObject::tr("bin file (*.bin)"));
+
+    // load data into QBytearray
+    QByteArray tmptile8x8data, tmptile8x8data_final;
+    QFile gfxbinfile(fileName);
+    if(!gfxbinfile.open(QIODevice::ReadOnly))
+    {
+        QMessageBox::critical(parent, QObject::tr("Error"), QObject::tr("Cannot open file! \n").append(gfxbinfile.errorString()));
+        return false;
+    }
+    if(!gfxbinfile.size())
+    {
+        QMessageBox::critical(parent, QObject::tr("Error"), QObject::tr("File size is 0!"));
+        return false;
+    }
+    tmptile8x8data = gfxbinfile.readAll();
+    tmptile8x8data_final = tmptile8x8data; // Init
+    gfxbinfile.close();
+
+    // Check size
+    if(tmptile8x8data.size() & 31)
+    {
+        QMessageBox::critical(parent, QObject::tr("Error"), QObject::tr("Illegal file size!\nIt should be a multiple of 32 Bytes."));
+        return false;
+    }
+
+    // Load palette data from bin file
+    fileName = QFileDialog::getOpenFileName(parent,
+                                            QObject::tr("Load palette bin file"), QString(""),
+                                            QObject::tr("bin file (*.bin)"));
+    QByteArray tmppalettedata;
+    QFile palbinfile(fileName);
+    if(!palbinfile.open(QIODevice::ReadOnly))
+    {
+        QMessageBox::critical(parent, QObject::tr("Error"), QObject::tr("Cannot open file!"));
+        return false;
+    }
+    tmppalettedata = palbinfile.readAll();
+    palbinfile.close();
+
+    QVector<QRgb> tmppalette;
+    unsigned short *tmppaldata = new unsigned short[16];
+    memset(tmppaldata, 0, 32);
+    memcpy(tmppaldata, tmppalettedata.data(), qMin(32, tmppalettedata.size()));
+    ROMUtils::LoadPalette(&tmppalette, tmppaldata, true);
+    delete[] tmppaldata;
+
+    // Get transparent color id in the palette
+    int transparentcolorId = 0;
+    SelectColorDialog scdialog;
+    scdialog.SetPalette(tmppalette);
+    scdialog.SetColor(0);
+    if(scdialog.exec() == QDialog::Accepted)
+    {
+        transparentcolorId = scdialog.GetSelectedColorId();
+    } else {
+        return false;
+    }
+
+    // transparent-substitute color replacement and load palette
+    tmppalette[transparentcolorId] = 0;
+
+    // nybble exchange not needed
+    // reset bytearray according to the palette bin file
+    for(int i = 0; i != 16; ++i)
+    {
+        char count = 0;
+
+        // Find if the color[i] is in the current palette
+        while(1)
```

Comment:
Can this whole `while(1)` be converted into a std::find_if ?

Reply:
just leave that there, i know the whole logic sucks, but i just hardly make that work last time, don't want to modify this part of code for now.